File: Services/java-services/Core/CoreDomain/src/main/java/ai/omnicure/core/domain/cache/CacheEntry.java
package ai.omnicure.core.domain.cache;

import ai.omnicure.core.shared.helper.CacheKeyHelper;
import lombok.Getter;
import lombok.Setter;

import java.lang.reflect.Constructor;

@Getter
@Setter
public abstract class CacheEntry<T> {

    protected final String prefix;
    private T value;

    protected CacheEntry(String prefix) {
        this.prefix = prefix;
    }

    public String getKey(String identifier) {
        return CacheKeyHelper.resolve(prefix, identifier);
    }

    public static <T, TEntry extends CacheEntry<T>> TEntry load(Class<TEntry> entryClass, String prefix, T value) {
        try {
            Constructor<TEntry> constructor = entryClass.getDeclaredConstructor(String.class);
            constructor.setAccessible(true);
            TEntry entry = constructor.newInstance(prefix);
            entry.setValue(value);
            return entry;
        } catch (Exception e) {
            throw new IllegalStateException("Cannot create CacheEntry instance of " + entryClass.getName(), e);
        }
    }

    public static <T, TEntry extends CacheEntry<T>> TEntry load(Class<TEntry> entryClass, String prefix) {
        return load(entryClass, prefix, null);
    }
}

File: Services/dotnet-services/Core/OmniCure.Core.Domain/Cache/CacheEntry.cs
namespace OmniCure.Core.Domain.Cache;

public abstract class CacheEntry<T>
{
    protected readonly string Prefix;
    public T? Value { get; set; }

    protected CacheEntry(string prefix)
    {
        Prefix = prefix;
    }

    public string GetKey(string identifier)
        => Shared.Helpers.CacheKeyHelper.Resolve(Prefix, identifier);

    public static TEntry Load<T2, TEntry>(string prefix, T2? value = default)
        where TEntry : CacheEntry<T2>
    {
        var entry = (TEntry)Activator.CreateInstance(typeof(TEntry), prefix)!;
        entry.Value = value;
        return entry;
    }
}
